server/integration/event-logger/src/main/java/org/infinispan/server/eventlogger/ServerEventLogger.java

```diff
@@ -89,26 +100,35 @@ TimeService getTimeService() {
    }
 
    @Override
-   public List<EventLog> getEvents(int start, int count, Optional<EventLogCategory> category, Optional<EventLogLevel> level) {
-      QueryBuilder query = getQueryFactory().from(ServerEventImpl.class).orderBy("when", SortOrder.DESC).maxResults(count).startOffset(start);
-      if (category.isPresent()) {
-         if (level.isPresent()) {
-            query.having("category").eq(category.get()).and().having("level").eq(level.get());
-         } else {
-            query.having("category").eq(category.get());
+   public List<EventLog> getEvents(Instant start, int count, Optional<EventLogCategory> category, Optional<EventLogLevel> level) {
+      List<EventLog> events = new ArrayList<>();
+      ConcurrentLinkedQueue<Throwable> throwables = new ConcurrentLinkedQueue<>();
+      try {
+         cacheManager.executor().submitConsumer(m -> {
+            Cache<Object, Object> cache = m.getCache(EVENT_LOG_CACHE);
+            QueryFactory<Query> queryFactory = Search.getQueryFactory(cache);
+            QueryBuilder query = queryFactory.from(ServerEventImpl.class).orderBy("when", SortOrder.DESC).maxResults(count);
+            FilterConditionContext filter = query.having("when").lte(start);
+            category.map(c -> filter.and(queryFactory.having("category").eq(c))).orElse(filter);
+            level.map(l -> filter.and(queryFactory.having("level").eq(l))).orElse(filter);
+            List<EventLog> nodeEvents = filter.toBuilder().build().list();
+            return nodeEvents;
+         }, (address, nodeEvents, t) -> {
+            if (t == null) {
+               events.addAll(nodeEvents);
+            } else {
+               throwables.add(t);
+            }
+         }).get(1, TimeUnit.MINUTES);
+         if (throwables.size() > 0) {
+            throw new CacheException(throwables.iterator().next());
```

Comment:
Oh wow, sorry I overlooked this.  You could just an AtomicReference instead and just take the last one instead even?  Then there is no need for a Collection.  But either way is fine with me.